tiledb/sm/filesystem/s3/STSCredentialsProvider.cc

```diff
@@ -0,0 +1,229 @@
+    : m_initialized(false) {
+  // check environment variables
+  Aws::String tmpRegion = Aws::Environment::GetEnv("AWS_DEFAULT_REGION");
+  m_roleArn = Aws::Environment::GetEnv("AWS_ROLE_ARN");
+  m_tokenFile = Aws::Environment::GetEnv("AWS_WEB_IDENTITY_TOKEN_FILE");
+  m_sessionName = Aws::Environment::GetEnv("AWS_ROLE_SESSION_NAME");
+
+  // check profile_config if either m_roleArn or m_tokenFile is not loaded from
+  // environment variable region source is not enforced, but we need it to
+  // construct sts endpoint, if we can't find from environment, we should check
+  // if it's set in config file.
+  if (m_roleArn.empty() || m_tokenFile.empty() || tmpRegion.empty()) {
+    auto profile =
+        Aws::Config::GetCachedConfigProfile(Aws::Auth::GetConfigProfileName());
+    if (tmpRegion.empty()) {
+      tmpRegion = profile.GetRegion();
+    }
+    // If either of these two were not found from environment, use whatever
+    // found for all three in config file
+    if (m_roleArn.empty() || m_tokenFile.empty()) {
+      m_roleArn = profile.GetRoleArn();
+      m_tokenFile = profile.GetValue("web_identity_token_file");
+      m_sessionName = profile.GetValue("role_session_name");
+    }
+  }
+
+  if (m_tokenFile.empty()) {
+    AWS_LOGSTREAM_WARN(
+        STS_ASSUME_ROLE_WEB_IDENTITY_LOG_TAG,
+        "Token file must be specified to use STS AssumeRole web identity creds "
+        "provider.");
+    return;  // No need to do further constructing
+  } else {
+    AWS_LOGSTREAM_DEBUG(
+        STS_ASSUME_ROLE_WEB_IDENTITY_LOG_TAG,
+        "Resolved token_file from profile_config or environment variable to be "
+            << m_tokenFile);
+  }
+
+  if (m_roleArn.empty()) {
+    AWS_LOGSTREAM_WARN(
+        STS_ASSUME_ROLE_WEB_IDENTITY_LOG_TAG,
+        "RoleArn must be specified to use STS AssumeRole web identity creds "
+        "provider.");
+    return;  // No need to do further constructing
+  } else {
+    AWS_LOGSTREAM_DEBUG(
+        STS_ASSUME_ROLE_WEB_IDENTITY_LOG_TAG,
+        "Resolved role_arn from profile_config or environment variable to be "
+            << m_roleArn);
+  }
+
+  if (tmpRegion.empty()) {
+    tmpRegion = Aws::Region::US_EAST_1;
+  } else {
+    AWS_LOGSTREAM_DEBUG(
+        STS_ASSUME_ROLE_WEB_IDENTITY_LOG_TAG,
+        "Resolved region from profile_config or environment variable to be "
+            << tmpRegion);
+  }
+
+  if (m_sessionName.empty()) {
+    m_sessionName = Aws::Utils::UUID::PseudoRandomUUID();
+  } else {
+    AWS_LOGSTREAM_DEBUG(
+        STS_ASSUME_ROLE_WEB_IDENTITY_LOG_TAG,
+        "Resolved session_name from profile_config or environment variable to "
+        "be "
+            << m_sessionName);
+  }
+
+  config.scheme = Aws::Http::Scheme::HTTPS;
+  config.region = tmpRegion;
+
+  if (!config.retryStrategy) {
+    Aws::Vector<Aws::String> retryableErrors;
+    retryableErrors.push_back("IDPCommunicationError");
+    retryableErrors.push_back("InvalidIdentityToken");
```

Comment:
The default retry strategies for the credentials providers retry some errors that our [S3-specific strategy](https://github.com/TileDB-Inc/TileDB/blob/5cee0cad1836d6a42ed39eb5a50875b07276d8fa/tiledb/sm/filesystem/s3.h#L1059) does not. How about I use a dedicated retry strategy (in a copy of the client config) for the credentials providers?

Reply:
I thatvis a good idea to use a separate retry strategy.

Reply:
Done.